### PR TITLE
Adds support for building on Solaris-derived systems

### PR DESCRIPTION
**Pull request by bnaecker**
This commit adds a number of changes to platform-detection and
compile-time directives required to support building ClickHouse on
Solaris-derived systems, most notably illumos.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Adds support for building on Solaris-derived operating systems.

Detailed description / Documentation draft:
This commit adds support for building on Solaris-derived systems, most notably [illumos](https://illumos.org). Most changes are non-functional, consisting of compile-time directives, inclusions, and defines in order to get the existing code to build. A few changes are to variable declarations or names, and are done to satisfy Solaris's include file structure and contents.

### CODE REVIEW

File: src/Common/Config/ConfigProcessor.cpp [hunk 1]
```diff
@@ -234,7 +234,7 @@ void ConfigProcessor::merge(XMLDocumentPtr config, XMLDocumentPtr with)
 
 static std::string layerFromHost()
 {
-    utsname buf;
+    struct utsname buf;
```
Comment (bnaecker) [inline src/Common/Config/ConfigProcessor.cpp:237]: I am actually curious how this code compiles under other platforms. The type `struct utsname` is almost universally _not_ typedef'd to just `utsname`.

File: src/Common/Stopwatch.cpp [hunk 2]
```diff
@@ -7,8 +7,12 @@ StopwatchRUsage::Timestamp StopwatchRUsage::Timestamp::current()
 
     ::rusage rusage {};
 #if !defined(__APPLE__)
+#if defined(OS_SUNOS)
+    ::getrusage(RUSAGE_LWP, &rusage);
```
Comment (bnaecker) [inline src/Common/Stopwatch.cpp:11]: For reference, LWP refers to a "light-weight process", and is functionally analogous to a thread on modern illumos systems.

File: utils/memcpy-bench/CMakeLists.txt [hunk 3]
```diff
@@ -15,6 +15,10 @@ add_executable (memcpy-bench
 
 add_compile_options(memcpy-bench PRIVATE -fno-tree-loop-distribute-patterns)
 
+if (OS_SUNOS)
+	target_compile_options(memcpy-bench PRIVATE "-Wa,--divide")
```
Comment (bnaecker) [inline utils/memcpy-bench/CMakeLists.txt:19]: The implementation of the memcpy assembly uses a `/` character, which on illumos's assembler is interpreted as the start of a comment, for compatibility with older assemblers.

File: src/Common/OpenTelemetryTraceContext.h [hunk 4]
```diff
@@ -11,7 +11,12 @@ struct OpenTelemetryTraceContext
     // The incoming tracestate header and the trace flags, we just pass them
     // downstream. See https://www.w3.org/TR/trace-context/
     String tracestate;
+
+#if defined(OS_SUNOS)
+    uint8_t trace_flags = 0;
+#else
```
Comment (alexey-milovidov) [inline src/Common/OpenTelemetryTraceContext.h:17]: This is probably by mistake, let's try to do this change for all platforms.

Comment (bnaecker) [reply src/Common/OpenTelemetryTraceContext.h:17]: Sounds good, I'll add a commit.

**Comment by alexey-milovidov** [comment]
A small typo:
```
2021-04-29 19:23:32 /ClickHouse/base/pcg-random/pcg_random.hpp:1646:18: error: use of undeclared identifier 'pgc_detail'; did you mean 'pcg_detail'?
2021-04-29 19:23:32 using ext_std8 = pgc_detail::extended<table_pow2, advance_pow2, BaseRNG,
2021-04-29 19:23:32                  ^~~~~~~~~~
2021-04-29 19:23:32                  pcg_detail
```

**Comment by bnaecker** [comment]
Good catch, thank you.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@bnaecker I heard that Illumos has Linux compatibility layer and can run Linux binaries, https://www.youtube.com/watch?v=TrfD3pC0VSs is that true?

**Comment by bnaecker** [comment]
@alexey-milovidov Yes, that's right! Zones are a general tool on illumos for isolating runtime environments and resources, and LX zones allow running a Linux kernel without a full VM. It works by translating the system calls, and is a convenient way to get Linux-only applications running. It does have some drawbacks though, which is minimal support for lots of the nice features of illumos itself related to debugging, ZFS, and service management.

**Comment by alexey-milovidov** [comment]
@bnaecker Also we need a pull request that updates every affected submodule to commits with changes.

**Comment by alexey-milovidov** [comment]
Ok. I've missed the notifications somehow.

**Comment by bnaecker** [comment]
Also, I'm not sure anything needs to be updated in `.gitmodules`, but I'd like to check. Of the patches that I've put up, only the modules referring to gRPC and Poco point to a specific branch. The rest pull from HEAD, if I'm not mistaken. The branches recorded in `.gitmodules` for those are `1.33.2` and `clickhouse`, respectively, both of which contain the commits merging my PRs.

So it seems no PR against ClickHouse itself should be required. Of course, please let me know if that's incorrect! Thanks.

**Comment by bnaecker** [comment]
@alexey-milovidov Ahh, disregard my earlier comment. I haven't worked with submodules much, and I see now that the changes in each have to be merged so that the superproject records a new commit ID as the submodule's commit. I'll work on that now, and put up a new PR with the required changes. Thanks!

**Comment by bnaecker** [comment]
@alexey-milovidov While updating the submodules, I've also discovered that one of my earlier PRs was incorrect, missing an `#endif` somewhere. I've put up [a new PR](https://github.com/ClickHouse-Extras/rocksdb/pull/7) with the fix, and verified it builds correctly on Linux and illumos. Apologies for that mistake!

**Pull request merged**